### ISSUE TITLE
Check for messages during derivative generation

### ISSUE THREAD

**Issue by mjordan** [SFULibrary/islandora_datastream_crud#76]
As pointed out by @DiegoPino in https://groups.google.com/d/msgid/islandora/c8202a7a-dead-496f-881f-06da74234dfc%40googlegroups.com?utm_medium=email&utm_source=footer, we should be checking for the existence of `$ds_message` and its member keys.

Thanks Heather for finding this and thanks Diego for suggesting the improvement!

**Comment by t4k** [comment]
I just came across the underlying issue here, as well.

I applied the changes you posted here, but still was getting errors.

> WD php: TypeError: Argument 2 passed to t() must be of the type array, null given, called in /usr/local/src/drush/includes/output.inc on line 148 in t() (line 1748 of    [error]
/…/includes/bootstrap.inc).
Cannot modify header information - headers already sent by (output started at /usr/local/src/drush/includes/output.inc:38) bootstrap.inc:1499                             [warning]
TypeError: Argument 2 passed to t() must be of the type array, null given, called in /usr/local/src/drush/includes/output.inc on line 148 in t() (line 1748 of /…/includes/bootstrap.inc).
Drush command terminated abnormally due to an unrecoverable error.                                                                                                        [error]

Just to see if I could address this, I changed the `$ds_message['message_sub']` argument to `array($ds_message['message_sub'])` and it seems to work. I'm getting "Created … derivative." messages printed and the script is not causing a Drush termination.

**Comment by mjordan** [comment]
@t4k thanks for the fix, I'll update that branch and ping you for some smoketests if that's OK.

**Comment by t4k** [comment]
I'll be happy to test.

**Comment by mjordan** [comment]
@t4k Strange, running in the `issue-76` branch, at 6e44240cf13b5322ff945529e3ef41c0b99c7943, I can't replicate those errors:

```
vagrant@islandora:/var/www/drupal/sites/all/modules/islandora_datastream_crud$ sudo drush islandora_datastream_crud_generate_derivatives --user=admin --source_dsid=OBJ --pid_file=/tmp/issue76.pids
You are about to regenerate derivatives from the OBJ datastream for 2 object(s) in your repository. This action cannot be undone. Do you want to want to continue? (y/n): y
Please be patient, generating derivatives from the OBJ datastream for islandora:13                                                                               [ok]
log4j:WARN No appenders could be found for logger (edu.harvard.hul.ois.fits.Fits).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
Created TECHMD derivative for (islandora:13).                                                                                                                    [ok]
The TN datastream was added successfully for islandora:13!                                                                                                       [ok]
The MEDIUM_SIZE datastream was added successfully for islandora:13!                                                                                              [ok]
Please be patient, generating derivatives from the OBJ datastream for islandora:14                                                                               [ok]
log4j:WARN No appenders could be found for logger (edu.harvard.hul.ois.fits.Fits).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
SLF4J: The following set of substitute loggers may have been accessed
SLF4J: during the initialization phase. Logging calls during this
SLF4J: phase were not honored. However, subsequent logging calls to these
SLF4J: loggers will work as normally expected.
SLF4J: See also http://www.slf4j.org/codes.html#substituteLogger
SLF4J: org.apache.tika.parser.ocr.TesseractOCRParser
Created TECHMD derivative for (islandora:14).                                                                                                                    [ok]
The TN datastream was added successfully for islandora:14!                                                                                                       [ok]
The MEDIUM_SIZE datastream was added successfully for islandora:14!
```
And nothing in my watchdog. I wrapped `$ds_message['message_sub']` in `array()` to replicate your fix and while the derivatives generated successfully, I'm now getting warnings containing uninterpolated `@pid`s:

```
vagrant@islandora:/var/www/drupal/sites/all/modules/islandora_datastream_crud$ sudo drush islandora_datastream_crud_generate_derivatives --user=admin --source_dsid=OBJ --pid_file=/tmp/issue76.pids
You are about to regenerate derivatives from the OBJ datastream for 2 object(s) in your repository. This action cannot be undone. Do you want to want to continue? (y/n): y
Please be patient, generating derivatives from the OBJ datastream for islandora:15                                                                               [ok]
log4j:WARN No appenders could be found for logger (edu.harvard.hul.ois.fits.Fits).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
htmlspecialchars() expects parameter 1 to be string, array given bootstrap.inc:1856                                                                              [warning]
Created @techmd derivative for (@pid).                                                                                                                           [ok]
htmlspecialchars() expects parameter 1 to be string, array given bootstrap.inc:1856                                                                              [warning]
The TN datastream was added successfully for @pid!                                                                                                               [ok]
htmlspecialchars() expects parameter 1 to be string, array given bootstrap.inc:1856                                                                              [warning]
The MEDIUM_SIZE datastream was added successfully for @pid!                                                                                                      [ok]
Please be patient, generating derivatives from the OBJ datastream for islandora:16                                                                               [ok]
log4j:WARN No appenders could be found for logger (edu.harvard.hul.ois.fits.Fits).
log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.
SLF4J: The following set of substitute loggers may have been accessed
SLF4J: during the initialization phase. Logging calls during this
SLF4J: phase were not honored. However, subsequent logging calls to these
SLF4J: loggers will work as normally expected.
SLF4J: See also http://www.slf4j.org/codes.html#substituteLogger
SLF4J: org.apache.tika.parser.ocr.TesseractOCRParser
htmlspecialchars() expects parameter 1 to be string, array given bootstrap.inc:1856                                                                              [warning]
Created @techmd derivative for (@pid).                                                                                                                           [ok]
htmlspecialchars() expects parameter 1 to be string, array given bootstrap.inc:1856                                                                              [warning]
The TN datastream was added successfully for @pid!                                                                                                               [ok]
htmlspecialchars() expects parameter 1 to be string, array given bootstrap.inc:1856                                                                              [warning]
The MEDIUM_SIZE datastream was added successfully for @pid!                                
```

According to https://api.drupal.org/api/drupal/includes%21bootstrap.inc/function/t/7.x, `t()` needs an associative array as its second parameter, so I am not understanding your `array()` fix. Would you mind posting your code here or in a gist so I can take a look at what you did?

**Comment by t4k** [comment]
Ah, here is my environment (just noticing I am on an old version of Drush, that may be an issue.):

Drush: 7.0.0
Drupal: 7.67
PHP: 7.1.18
ImageMagick 7.0.8

**Comment by mjordan** [comment]
I'm running on the Islandora Vagant, which has Drush version 6.3.0. I'll try to update to see if I can replicate what you're seeing.

**Comment by t4k** [comment]
I updated my Drush version to 8.3.1 and was still encountering the same issues.

This is what was happening before…
```
me@machine:~$ drush --user=1 --root=/var/www/current/docroot islandora_datastream_crud_generate_derivatives --source_dsid=OBJ --pid_file=/tmp/pids -v
Executing: mysql --defaults-extra-file=/tmp/drush_b0doAQ --database=islandora_db --host=db.example.com --silent  < /tmp/drush_2QxN45
Executing: mysql --defaults-extra-file=/tmp/drush_dg3tPl --database=islandora_db --host=db.example.com --silent  < /tmp/drush_0qAVcB
You are about to regenerate derivatives from the OBJ datastream for 1373 object(s) in your repository. This action cannot be undone. Do you want to want to continue? (y/n): y
Please be patient, generating derivatives from the OBJ datastream for image:5779                                                                                          [ok]
WD file: The file temporary://image-5779_OBJ.jp2 was not deleted, because it does not exist.                                                                                 [notice]
Undefined index: message_sub islandora_datastream_crud.drush.inc:592                                                                                                         [notice]
Drush command terminated abnormally due to an unrecoverable error.                                                                                                        [error]
```

I updated the `.drush.inc` file by changing the `||` to `&&` and adding an `elseif` condition…
```
      if ($message['success']) {
        foreach ($message['messages'] as $ds_message) {
          if (isset($ds_message) && (array_key_exists('message', $ds_message) && array_key_exists('message_sub', $ds_message))) {
            drush_log(strip_tags(dt($ds_message['message'], $ds_message['message_sub'])), 'ok');
          }
          elseif (isset($ds_message) && (array_key_exists('message', $ds_message) && !array_key_exists('message_sub', $ds_message))) {
            drush_log(strip_tags(dt($ds_message['message'])), 'ok');
          }
          else {
            drush_log(dt("Derivative generated, but no further information is available."), 'ok');
          }
        }
      }
      else {
        foreach ($message['messages'] as $ds_message) {
          if (isset($ds_message) && (array_key_exists('message', $ds_message) && array_key_exists('message_sub', $ds_message))) {
            drush_set_error('DERIVATIVE_NOT_GENERATED', strip_tags(dt($ds_message['message'], $ds_message['message_sub'])));
          }
          elseif (isset($ds_message) && (array_key_exists('message', $ds_message) && !array_key_exists('message_sub', $ds_message))) {
            drush_set_error('DERIVATIVE_NOT_GENERATED', strip_tags(dt($ds_message['message'])));
          }
          else {
            drush_set_error('DERIVATIVE_NOT_GENERATED', dt("Derivative not generated, no further information is available."));
          }
        }
      }
```

After that I am able to run the command…
```
me@machine:~$ drush --user=1 --root=/var/www/current/docroot islandora_datastream_crud_generate_derivatives --source_dsid=OBJ --pid_file=/tmp/pids -v
Executing: mysql --defaults-extra-file=/tmp/drush_b0doAQ --database=islandora_db --host=db.example.com --silent  < /tmp/drush_2QxN45
Executing: mysql --defaults-extra-file=/tmp/drush_dg3tPl --database=islandora_db --host=db.example.com --silent  < /tmp/drush_0qAVcB
You are about to regenerate derivatives from the OBJ datastream for 1372 object(s) in your repository. This action cannot be undone. Do you want to want to continue? (y/n): y
Please be patient, generating derivatives from the OBJ datastream for image:5775                                                                                          [ok]
WD file: The file temporary://image-5775_OBJ.jp2 was not deleted, because it does not exist.                                                                                 [notice]
Created TN derivative.                                                                                                                                                    [ok]
Created JPG derivative.                                                                                                                                                   [ok]
Created JP2 derivative.                                                                                                                                                   [ok]
Please be patient, generating derivatives from the OBJ datastream for image:5777                                                                                          [ok]
WD file: The file temporary://image-5777_OBJ.jp2 was not deleted, because it does not exist.                                                                                 [notice]
Created TN derivative.                                                                                                                                                    [ok]
Created JPG derivative.                                                                                                                                                   [ok]
Created JP2 derivative.                                                                                                                                                   [ok]
```

It may be difficult to reproduce my scenario because it might be arising due to the temporary file notice that is going on…

**Comment by DiegoPino** [comment]
@mjordan i just had the same issue. Seems like when derivative generation does not fail, but can not be done because source dsid is not there, the sucess message has missing parts. The fix was quite simple. 

Original:
```PHP
 // Print the output from the derivative generation functions.
    foreach ($messages as $message) {
      if ($message['success']) {
        foreach ($message['messages'] as $ds_message) {
          drush_log(strip_tags(dt($ds_message['message'], $ds_message['message_sub'])), 'ok');
        }
      }
      else {
        foreach ($message['messages'] as $ds_message) {
          drush_set_error('DERIVATIVE_NOT_GENERATED', strip_tags(dt($ds_message['message'], $ds_message['message_sub'])));
        }
      }
    }
````
```PHP
 // Print the output from the derivative generation functions.
    foreach ($messages as $message) {
      if ($message['success']) {
        foreach ($message['messages'] as $ds_message) {
          drush_log(strip_tags(dt($ds_message['message'], (array) $ds_message['message_sub'])), 'ok');
        }
      }
      else {
        foreach ($message['messages'] as $ds_message) {
          drush_set_error('DERIVATIVE_NOT_GENERATED', strip_tags(dt($ds_message['message'], (array) $ds_message['message_sub'])));
        }
      }
    }
````
In case this helps anybody, its just a simple cast to array.

**Comment by mjordan** [comment]
@DiegoPino thanks for the casting fix. @t4k does that work for you?

**Comment by DiegoPino** [comment]
Np! This module is a life saver ❤️ =)

**Comment by t4k** [comment]
@mjordan, yes I just successfully regenerated derivatives with the fix from @DiegoPino !

**Comment by mjordan** [comment]
Sweet - Occam's Razor at its best. Since we're all happy with that simple fix, I'll add it.

**Comment by mjordan** [comment]
Resolved with 3929dbf9ff16717a8e98f8ddfb2c38870949c1fd. Thanks for your help and patience @t4k , and thanks for the simple fix @DiegoPino.
